### DISCUSSION TITLE
Store CdsServiceClient between application restarts?

### DISCUSSION THREAD

**Discussion by kirilvalev** [microsoft/PowerPlatform-DataverseServiceClient#90] [Q&A] [answered]
Is it possible to store the CdsServiceClient so the user does not need to authenticate every time he starts the application?

Using the following code I am able to connect:

```
string url = "https://XXX.crm.dynamics.com";
string AppId = "0000000-0000000-000000-0000000";
string RedirectURI = "http://localhost";

string connectionString = $@"AuthType={AuthType};url={url};AppId={AppId};RedirectURI={RedirectURI}";

using var cdsClient = new CdsServiceClient(connectionString);
```

**Comment by sergeytunnik** [comment, marked as answer]
You can serialize token and store it in some kind of persistent storage (file, DB, etc).
Next initialize new CdsServiceClient instance with injecting token provider function which will deserialize it:
`public CdsServiceClient(Uri instanceUrl, Func<string, Task<string>> tokenProviderFunction, bool useUniqueInstance = true)`

Anyway token has lifetime which usually less than 24h. So if I logged in successfully today. Tomorrow by the time I will relaunch the app, the token will expire and I would have to enter credentials anyway.

**Reply by kirilvalev** [reply]
Thank you, didn't knew the token expires after 24 hours.
